scripts/builtin/shortestPath.dml

```diff
@@ -0,0 +1,157 @@
+# Grzegorz Malewicz, Matthew H. Austern, Aart J. C. Bilk, 
+# James C. Dehnert, Ikkan Horn, Naty Leiser and Grzegorz Czajkowski
+#
+#
+#
+#----------------------------------
+# *******INPUT PARAMETERS*********
+#----------------------------------
+#
+#	- Adjacency matrix of the labeled graph 
+#	(also considered directed labeled graphs)
+# 	
+#----------------------------------
+# *******OUTPUT PARAMETERS*********
+#----------------------------------
+#	- Matrix of minimum distances (shortest-path) between vertices
+#	(the value of the ith row and the jth column of the output matrix is 
+#	the minimum distance shortest-path from vertex i to vertex j)
+#	- When the value of the minimum distance is -1, the two nodes are not
+#	connected.
+#
+#----------------------------------
+# *******INTERMEDIATE ELEMENTS*********
+#----------------------------------
+#
+#	- neighboursList: matrix (double) of nx3 dimensions, when the minimum 
+#	distance from the source node to a given node is updated,the following
+#	information will be contained in this matrix. Each row represents one 
+#	different update.
+#		- First column: predecesor node which information about the minimum 
+#		distance from the source node has been updated.
+#		- Second column: neighbour that will receive the information in the 
+#		current superstep.
+#		- Third column: updated minimum distance from the source node to the
+#		 node in the first column.
+#	- neighboursListTmp: matrix (double) of nx3 dimensions, during each superstep,
+#	the information about the new updates (not needed until the following superstep)
+#	will be saved in this matrix.
+#
+#
+#
+
+
+s_shortestPath = function(Matrix[Double] G) 
+  return (Matrix[Double] C) 
+{
+	
+	print("SHORTEST PATH CALCULATION");
+
+	if(min(G) < 0){
+		stop("All values in G must be positive")
+	}
+	
+	if(ncol(G) != nrow(G)){
+		stop("Not correct matrix dimensions")
+	}
+	
+	matrixSize = nrow(G)
+	infValue = sum(rowSums(G)) + 1 # value representing infinity, i.e. the nodes are not connected
+	
+	# initialize the matrix of minimum distances with "infinity" values:
+	minDistMatrix = matrix(infValue,rows=matrixSize,cols=matrixSize)
+	
+	for (sourceNode in 1:matrixSize){ 
+		
+		# update minimum distance from the sourceNode to itself to 0:
+		minDistMatrix[sourceNode,sourceNode] = 0
+		
+		# initialize the matrix neighboursList and neighboursListTmp:
+		neighboursList  = matrix(0,matrixSize,3)
+		neighboursListTmp = matrix(0,matrixSize*matrixSize,3)
+		
+		# find the neighbours of the sourceNode and fill in the neighboursList:
+		nodeIdx = 1
+		for(ineighbour in 1:matrixSize){
+			if (as.integer(as.scalar(G[sourceNode,ineighbour]))>0 & (ineighbour!=sourceNode)){ 
+				neighboursList[nodeIdx,1] = sourceNode
+				neighboursList[nodeIdx,2] = ineighbour
+				nodeIdx = nodeIdx+1
+			}
+		}
+		
+		
+		while( as.integer(as.scalar(neighboursList[1,1])) > 0 ){ # loop of supersteps (see documentation)
```

Comment:
try to replace this complex loop with the loop from components() and modify this algorithm accordingly. Instead of propagating the max id in connected components we simply can take the current min distance of neighbors plus the distance of each node. This way the entire algorithms because a small few line vectorized loop.